src/peg/simulate.rs

```diff
@@ -1,70 +1,94 @@
 //! This module lets you interpret a PEG.
 
 use crate::cfg::Identifier;
-use crate::peg::{PegBody, PegFunction, PegProgram};
+use crate::peg::{PegBody, PegProgram};
 use crate::rvsdg::Expr;
 use bril_rs::{ConstOps, Literal, ValueOps};
+use std::cell::RefCell;
 use std::collections::HashMap;
+use std::rc::Rc;
 
-#[derive(Default)]
-struct Indices(HashMap<usize, usize>);
+#[derive(Clone, PartialEq, Eq, Hash)]
+enum Indices {
+    Root,
+    Node {
+        key: usize,
+        value: usize,
+        parent: Rc<Indices>,
+    },
+}
 
-impl Indices {
-    fn get(&self, label: usize) -> usize {
-        *self.0.get(&label).unwrap_or(&0)
+fn indices_get(indices: &Indices, label: usize) -> usize {
+    match indices {
+        Indices::Root => 0,
+        Indices::Node { key, value, .. } if *key == label => *value,
+        Indices::Node { parent, .. } => indices_get(parent, label),
     }
+}
 
-    fn set(&self, label: usize, value: usize) -> Indices {
-        let mut out = Indices(self.0.clone());
-        out.0.insert(label, value);
-        out
-    }
+fn indices_set(indices: Rc<Indices>, key: usize, value: usize) -> Rc<Indices> {
+    Rc::new(Indices::Node {
+        key,
+        value,
+        parent: indices,
+    })
 }
 
 impl PegProgram {
     pub fn simulate(&self, args: &[Literal]) -> String {
-        let mut stdout = String::new();
-        let main = self.functions.iter().find(|f| f.name == "main").unwrap();
-        let output = main.simulate(args, self, &mut stdout);
+        let main = self
+            .functions
+            .iter()
+            .position(|f| f.name == "main")
+            .unwrap();
+        let mut s = Simulator {
+            args: args.to_vec(),
+            indices: Rc::new(Indices::Root),
+            func: usize::MAX, // garbage value
+            program: self,
+            stdout: Rc::default(),
+            memoizer: Rc::default(),
+        };
+        let output = s.simulate_func(main);
         assert!(output.is_none());
-        stdout
+        Rc::try_unwrap(s.stdout).unwrap().into_inner()
     }
 }
 
-impl PegFunction {
-    pub fn simulate(
-        &self,
-        args: &[Literal],
-        program: &PegProgram,
-        stdout: &mut String,
-    ) -> Option<Literal> {
-        assert_eq!(self.n_args, args.len());
-        assert!(self.nodes[self.state]
-            .simulate(args, &self.nodes, &Indices::default(), program, stdout)
-            .is_none());
-        self.result.and_then(|body| {
-            self.nodes[body].simulate(args, &self.nodes, &Indices::default(), program, stdout)
-        })
-    }
+#[derive(Clone)]
+struct Simulator<'a> {
+    args: Vec<Literal>,
+    indices: Rc<Indices>,
+    func: usize,
+    program: &'a PegProgram,
+    stdout: Rc<RefCell<String>>,
+    memoizer: Rc<RefCell<Memoizer>>,
 }
 
-impl PegBody {
+type Memoizer = HashMap<(usize, usize, Rc<Indices>), Literal>;
```

Comment:
what are these usize arguments? I think this and Indices could use some documentation